portal-impl/src/main.rs

```diff
@@ -0,0 +1,228 @@
+#[doc(hidden)]
+enum ResponseType {
+    Success = 0,
+    Cancelled = 1,
+    Other = 2,
+}
+
+#[derive(zbus::DBusError, Debug)]
+enum Error {
+    Msg(String),
+}
+
+impl Error {
+    fn new(msg: &str) -> Error {
+        Error::Msg(msg.to_string())
+    }
+}
+
+impl From<oo7::dbus::Error> for Error {
+    fn from(err: oo7::dbus::Error) -> Self {
+        Error::new(&err.to_string())
+    }
+}
+
+struct Secret;
+
+#[dbus_interface(name = "org.freedesktop.impl.portal.Secret")]
+impl Secret {
+    #[dbus_interface(property, name = "version")]
+    fn version(&self) -> u32 {
+        PORTAL_VERSION
+    }
+
+    #[dbus_interface(out_args("response", "results"))]
+    async fn retrieve_secret(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+        handle: OwnedObjectPath,
+        app_id: &str,
+        fd: zvariant::Fd,
+        _options: HashMap<&str, zvariant::Value<'_>>,
+    ) -> Result<(ResponseType, HashMap<&str, zvariant::OwnedValue>), Error> {
+        tracing::debug!("Got request from {app_id} with options: {_options:?}");
+
+        let (sender, receiver) = futures_channel::oneshot::channel();
+
+        if let Err(err) = Request::serve(object_server, handle.clone(), sender).await {
+            tracing::error!("Could not register object {handle}: {err}");
+        }
+
+        let fut_1 = async move {
+            let res = match retrieve_secret_inner(app_id, fd).await {
+                Ok(res) => Ok((ResponseType::Success, res)),
+                Err(err) => {
+                    tracing::error!("could not retrieve secret: {err}");
+                    Ok((ResponseType::Other, HashMap::new()))
+                }
+            };
+
+            // We do not accept Close request anymore here.
+            tracing::debug!("Object {handle} handled");
+            object_server
+                .remove::<Request, &zvariant::OwnedObjectPath>(&handle)
+                .await
+                .unwrap();
+
+            res
+        };
+
+        let fut_2 = async move {
+            receiver.await.unwrap();
+            Ok((ResponseType::Cancelled, HashMap::new()))
+        };
+
+        let t1 = fut_1.fuse();
+        let t2 = fut_2.fuse();
+
+        futures_util::pin_mut!(t1, t2);
+
+        futures_util::select! {
+            fut_1_res = t1 => fut_1_res,
+            fut_2_res = t2 => fut_2_res,
+        }
+    }
+}
+
+fn generate_secret() -> Result<zeroize::Zeroizing<Vec<u8>>, Error> {
+    let mut secret = [0; PORTAL_SECRET_SIZE];
+    let rand = ring::rand::SystemRandom::new();
+    rand.fill(&mut secret)
+        .map_err(|err| Error::new(&err.to_string()))?;
+    Ok(zeroize::Zeroizing::new(secret.to_vec()))
+}
+
+async fn retrieve_secret_inner(
+    app_id: &str,
+    fd: zvariant::Fd,
+) -> Result<HashMap<&'static str, zvariant::OwnedValue>, Error> {
+    let service = Service::new(Algorithm::Encrypted)
+        .await
+        .or(Service::new(Algorithm::Plain).await)?;
```

Comment:
```suggestion
    let service = Service::new(Algorithm::Encrypted)
        .await
        .or(Service::new(Algorithm::Plain).await)?;
```
should no longer be needed with https://github.com/bilelmoussaoui/oo7/pull/64

Reply:
Done.